packages/frontend2/src/server/features/scaling/tvl/breakdown/get-tvl-breakdown.ts

```diff
@@ -0,0 +1,170 @@
+            asset.amount += amountAsNumber
+            asset.escrows.push({
+              amount: amountAsNumber,
+              usdValue: valueAsNumber,
+              escrowAddress: config.escrowAddress,
+              ...(config.type === 'preminted' ? { isPreminted: true } : {}),
+            })
+          } else {
+            breakdown.canonical.set(priceConfig.assetId, {
+              assetId: priceConfig.assetId,
+              chainId: chainConverter.toChainId(config.chain),
+              amount: amountAsNumber,
+              usdValue: valueAsNumber,
+              usdPrice: price.toString(),
+              escrows: [
+                {
+                  amount: amountAsNumber,
+                  usdValue: valueAsNumber,
+                  escrowAddress: config.escrowAddress,
+                  ...(config.type === 'preminted' ? { isPreminted: true } : {}),
+                },
+              ],
+            })
+          }
+          break
+        }
+        case 'external': {
+          const token = safeGetTokenByAssetId(priceConfig.assetId)
+
+          breakdown.external.push({
+            assetId: priceConfig.assetId,
+            chainId: chainConverter.toChainId(config.chain),
+            amount: amountAsNumber,
+            usdValue: valueAsNumber,
+            usdPrice: price.toString(),
+            tokenAddress:
+              config.address === 'native' ? undefined : config.address,
+            bridge: config.bridge ?? {
+              name: token?.bridgedUsing?.bridge ?? 'Unknown',
+              slug: token?.bridgedUsing?.slug,
+              warning: token?.bridgedUsing?.warning,
+            },
+          })
+          break
+        }
+        case 'native':
+          breakdown.native.push({
+            assetId: priceConfig.assetId,
+            chainId: chainConverter.toChainId(config.chain),
+            amount: amountAsNumber,
+            usdValue: valueAsNumber,
+            usdPrice: price.toString(),
+            // TODO: force fe to accept "native"
+            tokenAddress:
+              config.address === 'native' ? undefined : config.address,
+          })
+      }
+    }
+    const result: {
+      canonical: CanonicalAssetBreakdownData[]
+      external: ExternalAssetBreakdownData[]
+      native: NativeAssetBreakdownData[]
+    } = {
+      canonical: [],
+      external: [],
+      native: [],
+    }
+
+    const canonical = [...breakdown.canonical.values()].sort(
+      (a, b) => +b.usdValue - +a.usdValue,
+    )
+
+    result.canonical = canonical.map((x) => ({
+      ...x,
+      escrows: x.escrows.sort((a, b) => +b.amount - +a.amount),
+      amount: x.amount,
+      usdValue: x.usdValue,
+    }))
+
+    result.external = breakdown.external.sort(
+      (a, b) => +b.usdValue - +a.usdValue,
+    )
+
+    result.native = breakdown.native.sort((a, b) => +b.usdValue - +a.usdValue)
+    return { dataTimestamp: targetTimestamp, breakdown: result }
+  }
+}
```

Comment:
it looks like we can break this down into separate functions

Reply:
https://github.com/l2beat/l2beat/pull/4855/commits/91577e2e829549645bcb8ffe3f796a447f3ef579